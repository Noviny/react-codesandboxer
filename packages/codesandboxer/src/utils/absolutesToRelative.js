// @flow

export default function absolutesToAbsolute(
  currentLocation: string,
  targetLocation: string,
) {
  let cr = currentLocation.split('/');
  let tr = targetLocation.split('/');

  // remove any shared values
  while (tr[0] && cr[0] === tr[0]) {
    cr.shift();
    tr.shift();
  }

  if (cr.length < 2) {
    return `./${tr.join('/')}`;
  }

  cr.shift();

  return `${cr.map(() => '..').join('/')}/${tr.join('/')}`;
}
