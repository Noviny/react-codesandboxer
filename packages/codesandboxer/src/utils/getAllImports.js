// @flow
import type { Import } from '../types';

let matchingDeps = '\\s*[\'"`]([^\'"`]+)[\'"`]\\s*';
let matchingName = '\\s*(?:[\\w${},\\s*]+)\\s*';
// the first half of this regex matches require statements
let regex =
  '(?:(?:var|const|let)' +
  matchingName +
  '=\\s*)?require\\(' +
  matchingDeps +
  '\\);?';

// the second half of this regex matches imports
regex +=
  '|(?:import|export)(?:' + matchingName + 'from\\s*)?' + matchingDeps + ';?';

const getAllImports = (code: string): Array<Import> => {
  let matcher = new RegExp(regex, 'g');
  let matches = [];
  let m = matcher.exec(code);

  while (m) {
    matcher = new RegExp(regex, 'g');
    matches.push(m[1] || m[2] || m[3] || m[4]);
    code = code.slice(m.index + m[0].length);
    m = matcher.exec(code);
  }
  return matches;
};

export default getAllImports;
