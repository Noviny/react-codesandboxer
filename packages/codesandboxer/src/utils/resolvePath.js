// @flow
const rp2 = (base, rel) => {
  switch (rel[0]) {
    case '..': {
      return rp2(base.slice(0, -1), rel.slice(1));
    }
    default: {
      return `${base.concat(rel).join('/')}`;
    }
  }
};

const resolvePath = (basePath: string, relativePath: string): string => {
  let base = basePath.split('/').filter(a => a);
  let rel = relativePath.split('/').filter(a => a);

  // if the relative path is too long, we won't be able to resolve it.
  if (rel.length > base.length + 1) {
    // TODO: Discuss what a good 'failcase' behaviour is for paths we can't resolve
    return basePath;
  }
  let val = '';
  switch (rel[0]) {
    case '.':
      val = base
        .slice(0, -1)
        .concat(rel.slice(1))
        .join('/');
      break;
    case '..':
      val = rp2(base.slice(0, -1), rel);
      break;
    default:
      // TODO: Decide if this is correct or terribly wrong
      val = base.concat(rel).join('/');
  }

  return val;
};

export default resolvePath;
