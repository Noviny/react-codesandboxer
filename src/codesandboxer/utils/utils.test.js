// @flow
import cases from 'jest-in-case';
import getAllImports from './getAllImports';
import resolvePath from './resolvePath';
import absolutesToRelative from './absolutesToRelative';
import path from 'path';

const codeImportTests = [
  { name: 'simple import', code: `import a from 'b'` },
  { name: 'spread import', code: `import { a } from 'b'` },
  { name: 'two imports', code: `import a from 'b' import c from 'd'` },
  {
    name: 'multiline imports',
    code: `import a from 'b'
import c from 'd'`,
  },
  { name: 'two spread imports', code: `import { a, b } from 'c'` },
  {
    name: 'two spread imports multiline',
    code: `import {
  a,
  b
} from 'c'`,
  },
  { name: 'no spaces', code: `import {a} from 'b'` },
  {
    name: 'dev and peer deps',
    code: `import {a} from 't' import s from 'z' import y from 'x'`,
  },
  {
    name: 'relativeImport',
    code: `import {a} from './c'`,
  },
  {
    name: 'using regex pattern',
    code: `import a from './c/somewhere' import b from './c/anywhere'`,
    old: './c/*',
    new: 'c/',
  },
  {
    name: 'import then immediately export',
    code: `export { default } from './abc'`,
    old: './abc',
    new: './cde',
  },
  {
    name: 'import then immediately export as value',
    code: `export { default as something } from './abc'`,
    old: './abc',
    new: './cde',
  },
  {
    name: 'import then immediately export not default',
    code: `export { urd as something } from './abc' } from 'esk'`,
    old: './abc',
    new: './cde',
  },
];

cases(
  'getAllImports()',
  ({ code }) => {
    let mpts = getAllImports(code);
    expect(mpts).toMatchSnapshot();
  },
  codeImportTests,
);

cases(
  'resolvePath()',
  ({ basePath, relativePath, returnedPath }) => {
    let res = resolvePath(basePath, relativePath);
    expect(res).toBe(returnedPath);
  },
  [
    { basePath: 'a/b/c', relativePath: '../z', returnedPath: 'a/z' },
    {
      basePath: 'a/b/c',
      relativePath: '../../z/x',
      returnedPath: 'z/x',
    },
    { basePath: 'a/b/c', relativePath: './z', returnedPath: 'a/b/z' },
    {
      basePath: 'a/b/c',
      relativePath: '../../../z/x',
      returnedPath: 'a/b/c',
    },
    {
      basePath: 'a/b/c',
      relativePath: 'zxy',
      returnedPath: 'a/b/c/zxy',
    },
    {
      basePath: 'a/b/c/',
      relativePath: './zxy',
      returnedPath: 'a/b/zxy',
    },
  ],
);

cases(
  'absolutesToRelative different dirs',
  ({ currentLocation, targetLocation }) => {
    let actualPath = path.relative(
      path.dirname(currentLocation),
      targetLocation,
    );
    expect(absolutesToRelative(currentLocation, targetLocation)).toBe(
      actualPath,
    );
  },
  [
    {
      name: 'basic case',
      currentLocation: 'examples/somewhere.js',
      targetLocation: 'src',
    },
    {
      name: 'more deeply nested',
      currentLocation: 'examples/deeper/somewhere.js',
      targetLocation: 'src',
    },
    {
      name: 'in same subdirectory',
      currentLocation: 'examples/deeper/somewhere.js',
      targetLocation: 'examples/fork/elsewhere',
    },
  ],
);

cases(
  'absolutesToRelative same dir',
  ({ currentLocation, targetLocation }) => {
    let actualPath = path.relative(
      path.dirname(currentLocation),
      targetLocation,
    );
    expect(absolutesToRelative(currentLocation, targetLocation)).toBe(
      `./${actualPath}`,
    );
  },
  [
    {
      name: 'in same directory',
      currentLocation: 'examples/deeper/somewhere.js',
      targetLocation: 'examples/deeper/elsewhere',
    },
    {
      name: 'in subdirectory',
      currentLocation: 'examples/somewhere.js',
      targetLocation: 'examples/deeper/elsewhere',
    },
  ],
);
