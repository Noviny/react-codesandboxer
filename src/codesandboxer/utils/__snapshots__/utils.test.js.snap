// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAllImports() dev and peer deps 1`] = `
Array [
  Array [
    "import {a} from 't'",
    "t",
  ],
  Array [
    "import s from 'z'",
    "z",
  ],
  Array [
    "import y from 'x'",
    "x",
  ],
]
`;

exports[`getAllImports() import then immediately export 1`] = `
Array [
  Array [
    "export { default } from './abc'",
    "./abc",
  ],
]
`;

exports[`getAllImports() import then immediately export as value 1`] = `
Array [
  Array [
    "export { default as something } from './abc'",
    "./abc",
  ],
]
`;

exports[`getAllImports() import then immediately export not default 1`] = `
Array [
  Array [
    "export { urd as something } from './abc'",
    "./abc",
  ],
]
`;

exports[`getAllImports() multiline imports 1`] = `
Array [
  Array [
    "import a from 'b'",
    "b",
  ],
  Array [
    "import c from 'd'",
    "d",
  ],
]
`;

exports[`getAllImports() no spaces 1`] = `
Array [
  Array [
    "import {a} from 'b'",
    "b",
  ],
]
`;

exports[`getAllImports() relativeImport 1`] = `
Array [
  Array [
    "import {a} from './c'",
    "./c",
  ],
]
`;

exports[`getAllImports() simple import 1`] = `
Array [
  Array [
    "import a from 'b'",
    "b",
  ],
]
`;

exports[`getAllImports() spread import 1`] = `
Array [
  Array [
    "import { a } from 'b'",
    "b",
  ],
]
`;

exports[`getAllImports() two imports 1`] = `
Array [
  Array [
    "import a from 'b'",
    "b",
  ],
  Array [
    "import c from 'd'",
    "d",
  ],
]
`;

exports[`getAllImports() two spread imports 1`] = `
Array [
  Array [
    "import { a, b } from 'c'",
    "c",
  ],
]
`;

exports[`getAllImports() two spread imports multiline 1`] = `
Array [
  Array [
    "import {
  a,
  b
} from 'c'",
    "c",
  ],
]
`;

exports[`getAllImports() using regex pattern 1`] = `
Array [
  Array [
    "import a from './c/somewhere'",
    "./c/somewhere",
  ],
  Array [
    "import b from './c/anywhere'",
    "./c/anywhere",
  ],
]
`;
